### PR TITLE
feat: add status message below prompt editor

### PR DESCRIPTION
**Pull request by audrey-kho**
## Description

- An error message is shown under the prompt editor when an unsupported character is detected
- Generate button is disabled when an unsupported character is used

## Related Issue

 
[https://jira.corp.adobe.com/browse/SITES-21736](https://jira.corp.adobe.com/browse/SITES-21736)

## How Has This Been Tested?
[https://experience-qa.adobe.com/?devMode=true&shell_source=workspace&workspace=charerrorui&shell_ims=prod#/aem/generate-variations/](https://experience-qa.adobe.com/?devMode=true&shell_source=workspace&workspace=charerrorui&shell_ims=prod#/aem/generate-variations/)

## Screenshots (if appropriate):
Screenshot 2024-05-29 at 9 16 44 AM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: web-src/src/components/PromptEditor.js [hunk 1]
```diff
@@ -196,6 +229,22 @@ function PromptEditor({ isOpen, onClose, ...props }) {
...
+            </Text>
+          </Flex>
+
+          <Flex gap="size-100" UNSAFE_className={editorSyntaxError ? style.errorHelpText : style.hidden}>
+            <Alert aria-label="Negative Alert" color="negative" />
+            <Text>
+              The characters <span style={{ fontWeight: '600' }}>&#123;</span>, <span style={{ fontWeight: '600' }}>&#125;</span>,
+              and <span style={{ fontWeight: '600' }}>&quot;</span> are reserved and can&apos;t be used within quoted text values.
+              Please remove or replace these characters and try again.
+            </Text>
```
Comment (askayastha22) [inline web-src/src/components/PromptEditor.js:246]: Localizable text. It's a good opportunity to localize a very complex text string.
Please refer to this example: https://github.com/adobe/aem-genai-assistant/blob/3c26d313fbea58515e8710e209144592c0a99356/web-src/src/components/ConsentDialog.js#L72

**Review by askayastha22 — CHANGES_REQUESTED** [review]

File: web-src/src/components/GenerateButton.js [hunk 2]
```diff
@@ -154,7 +156,7 @@ export function GenerateButton() {
         variant="cta"
         style="fill"
         onPress={handleGenerate}
-        isDisabled={generationInProgress}>
+        isDisabled={generationInProgress || syntaxError}>
```
Comment (slitviachenko) [inline web-src/src/components/GenerateButton.js:159]: I'll probably leave this to @vtsaplin and @askayastha22 since they know the architecture of this app better. However, having a new shared Recoil state to control the availability of this button is not the best option, in my opinion. What if later there is another condition, such as a requirement to provide a target audience? Will another Recoil state need to be created for that?

I understand that the `GenerateButton` is already too coupled and it is impossible to reuse it in other components. However, it would be better to address the control of its availability at a minimum. The logic for this button's availability should be handled by the parent component, using the `isDisabled` prop, for example.

File: web-src/src/components/PromptEditor.js [hunk 3]
```diff
@@ -196,6 +227,22 @@ function PromptEditor({ isOpen, onClose, ...props }) {
               readOnly={!viewSource}
             />
           </div>
+
+          <Flex gap="size-100" UNSAFE_className={editorSyntaxError ? style.hidden : style.validHelpText }>
+            <CheckmarkCircle aria-label="Positive Alert" color="positive" />
+            <Text>
+              Prompt is valid.
```
Comment (slitviachenko) [inline web-src/src/components/PromptEditor.js:234]: Usually, error messages are displayed only. Do we need to display this? CC: @lenaorobei

File: web-src/src/components/PromptSessionSideView.js [hunk 4]
```diff
@@ -116,7 +118,7 @@ export function PromptSessionSideView({ isOpenPromptEditor, onTogglePrompt, ...p
           <SavePromptButton />
           <ResetButton />
         </Flex>
-        <GenerateButton />
+        <GenerateButton disabled={disableTextGeneration} />
```
Comment (slitviachenko) [inline web-src/src/components/PromptSessionSideView.js:121]: Should the `SavePromptButton` above be disabled as well? 🤔 Otherwise, clients might store and share broken prompts with their organization.

**Review by askayastha22 — APPROVED** [review]
LGTM

File: web-src/src/components/PromptSessionPanel.js [hunk 5]
```diff
@@ -51,9 +53,13 @@ export function PromptSessionPanel() {
         }}>
           <PromptResultListView style={{ height: '100%' }} onClick={() => setIsOpenPromptEditor(false)} />
         </Flex>
-        <PromptEditor isOpen={isOpenPromptEditor} onClose={() => setIsOpenPromptEditor(false)} />
+        <PromptEditor
+          isOpen={isOpenPromptEditor}
+          onClose={() => setIsOpenPromptEditor(false)}
+          error={promptEditorError}
```
Comment (slitviachenko) [inline web-src/src/components/PromptSessionPanel.js:59]: From my understanding of the app's architecture, the `PromptSessionPanel` component might be considered a `form` component. It could have its own validation state, which other child components might call to notify about any validation errors. Child components like `PromptSessionSideView` and `PromptEditor` might have their own _internal_ validation states to track their validation logic.

```
PromptSessionPanel (a form element) - has an internal state to track form validation and form controls
  |-- PromptSessionSideView (an input element)
  |-- PromptEditor (an input element) - has its own internal state to track its validation logic and notifies PromptSessionPanel in case of errors
```
- once `PromptEditor` encounters a validation error, it should notify `PromptSessionPanel`
- in turn, `PromptSessionPanel` should notify `PromptSessionSideView` to disable `GenerateButton`

In short, passing `promptEditorError` isn't needed. `PromptEditor` should have its own state to control its validation logic and only notify `PromptSessionPanel` about errors by calling `PromptSessionPanel`'s `setPromptEditorError` method. However, we need to rename it to something like `setPromptValidationError` since other child components that we could have might also fail their own validation logic.

However, I need input from @askayastha22 as well.

**Review by slitviachenko — COMMENTED** [review]
@audrey-kho , I left some comments, could you please take a look when you have time?
CC: @askayastha22

File: web-src/src/components/PromptSessionSideView.js (hunk 4 above)
Comment (askayastha22) [reply web-src/src/components/PromptSessionSideView.js:121]: That's a good point.

File: web-src/src/components/PromptSessionPanel.js (hunk 5 above)
Comment (askayastha22) [reply web-src/src/components/PromptSessionPanel.js:59]: @slitviachenko 
- From my point of view, even though PromptSessionPanel looks like a form element, it's dynamically created via prompt syntax. So, validation logic for each text field is hard to determine beforehand.
- As for your point about `PromptEditor` having its own state to control validation logic, it is a good catch and we may refactor the code.

File: web-src/src/components/PromptSessionSideView.js [hunk 6]
```diff
@@ -61,11 +61,15 @@ const styles = {
...
 
-export function PromptSessionSideView({ isOpenPromptEditor, onTogglePrompt, ...props }) {
+export function PromptSessionSideView({
+  isOpenPromptEditor, onTogglePrompt, promptEditorError, ...props
+}) {
   const currentSession = useRecoilValue(sessionState);
   const [viewType, setViewType] = useRecoilState(viewTypeState);
   const { formatMessage } = useIntl();
 
+  console.log('PromptSessionSideView.js: promptEditorError:', promptEditorError);
```
Comment (slitviachenko) [inline web-src/src/components/PromptSessionSideView.js:71]: Please remove this `console.log`

**Review by slitviachenko — APPROVED** [review]
Please remove the `console.log` statement, the rest LGTM
